src/main/scala/components/memory-async.scala

```diff
@@ -0,0 +1,378 @@
+// Asynchronous memory module
+
+package dinocpu
+
+import chisel3._
+import chisel3.util._
+
+import chisel3.experimental.ChiselEnum
+
+import chisel3.util.experimental.loadMemoryFromFile
+import firrtl.annotations.MemoryLoadFileType
+
+// A Bundle used for representing a memory access by instruction memory or data memory.
+class Request extends Bundle {  
+  val address   = UInt(32.W)
+  val writedata = UInt(32.W)
+
+  // Encodes the operation requested by this access
+  //  0 -> nop
+  //  1 -> read
+  //  2 -> write
+  val operation = UInt(2.W)
+}
+
+// A Bundle used for temporarily storing the necessary information for a write in the data memory accessor.
+class PartialWrite extends Bundle {
+  val address   = UInt(32.W)
+  val writedata = UInt(32.W)
+  val maskmode  = UInt(2.W)
+
+  // Used to determine whether the stored write is actually a valid write.
+  val valid     = Bool() 
+}
+
+/** 
+ * The generic interface for communication between the IMem/DMemAccess modules and the backing memory.
+ *
+ * Input:  request, the ready/valid interface for a MemAccess module to issue Requests to. Memory
+ *         will only accept a request when both request.valid (the MemAccess is supplying valid data)
+ *         and request.ready (the memory is idling for a request) are high.
+ *
+ * Output: response, the valid interface for the data outputted by memory if it was requested to read. 
+ *         the bits in response.bits should only be treated as valid data when response.valid is high.
+ */
+class AsyncMemIO extends Bundle {
+  val request  = Flipped(Decoupled (new Request))
+  val response = Valid (UInt (32.W))
+}
+
+/** 
+ * A generic ready/valid interface for MemAccess modules, whose IOs extend this.
+ * Input:  address, the address of a piece of data in memory. 
+ * Input:  valid, true when the address specified is valid
+ * Input:  response, the return route from memory to a memory access. This is primarily meant for connecting to 
+ *         an AsyncMemIO's response output, and should not be connected to anything else in any circumstance 
+ *         (or things will possibly break)
+ * 
+ * Output: ready, true when memory is either idling or outputting a value, and is ready for a new 
+ *         request. Note that this is different from access_out.ready - this ready is for use by the general
+ *         CPU (like to signal when to stall the CPU), while access_out.ready is used for signalling 
+ *         between the memory and accessor only
+ * Output: request, a DecoupledIO that delivers a request from a memory accessor to memory. This is primarily
+ *         meant for connecting to an AsynMemIO's request input, and should not be connected to anything else
+ */
+class MemAccessIO extends Bundle {
+  val address  = Input(UInt(32.W))
+  val valid    = Input(Bool())
+  val response = Flipped(Valid(UInt(32.W)))
```

Comment:
Maybe we should group `address`, `valid`, and `ready` together since this is the "pipeline" interface, and `response` and `request` should be grouped together since they are the "memory side" interface.

This will make it a little clearer when you extend them below, I think.